Issue/11145 start ucrop with high res image

ashiagr opened the pull request:
Fixes #11145

This PR
- Starts `uCrop` from the preview fragment with the high res image file path.
- Returns to the original post on the `Gutenberg` editor skipping preview fragment on back action.

P.S.
Few observations related to `uCrop` fragment which may require separate tasks (adding them as  a checklist): 

- [X] On back action from the app bar on the `uCrop` fragment, we want to go to the original post on the `Gutenberg` editor skipping preview fragment. 
    Using `popUpToInclusive` property for this purpose  doesn't pop the start destination (preview fragment) off the back stack  in a module.
     On doing so, [activity finishes](https://android.googlesource.com/platform/frameworks/support/+/refs/heads/androidx-master-dev/navigation/navigation-runtime/src/main/java/androidx/navigation/NavController.java#386) and then a failed attempt is made to reopen the start destination. As a workaround, explicitly invoking back action for the `uCrop` fragment. Will further investigate it when we pass edited image to the `Gutenberg` editor.

    Related Android issue [here](https://issuetracker.google.com/issues/147312109).

    **UPDATE**: As discussed in comments, workaround looks ok. 

- [ ] Using temporary options (aspect ratios, crop frame/ grid) and styling as shown in this screenshot. Will update with actual options after a discussion with the designer.
cc @mbshakti 

![device-2020-02-25-140701](https://user-images.githubusercontent.com/1405144/75229746-8be7d400-57d8-11ea-8dd9-89224d150cb2.png)

- [ ] The cropping rectangle in the `uCrop` fragment is currently shown behind the toolbar. As a result we cannot drag upper corners. Propagating events to the view behind (by setting `clickable`, `focusable` to `false`) doesn't seem to be working in case of [Toolbar](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/widget/Toolbar.java#1585) as it eats all touch events, regardless of the `clickable` attribute and `@null` background.

```
@Override
public boolean onTouchEvent(MotionEvent ev) {
  // Toolbars always eat touch events
```
Will take care of this issue in a separate PR.

- [X] `uCrop` doesn't seem to be saving image state and restoring it on orientation change (at least with the version we're using`2.2.2`). Will investigate it more in a separate task. 

    Related `uCrop` issue [here](https://git.io/JvE7P).
    
   **UPDATE**: As discussed in comments, leaving this issue to the end of the project. Added a note in the Media Editing project.

- [X] Yet to decide output file path.

    **UPDATE**: As discussed in comments, output file location is ok.

- [X] Noticed a bug, low res(LR) image was saved to file and sent to `uCrop` on moving the app to background while LR image was loading and bringing the app to foreground after it was loaded but before HR image was loaded. 

    It is probably happening because `viewModel.onLoadIntoImageViewSuccess(url)` is called again due to the [LiveData fact](https://developer.android.com/topic/libraries/architecture/livedata#the_advantages_of_using_livedata):

    > If a lifecycle becomes inactive, it receives the latest data upon becoming active again. For example, an activity that was in the background receives the latest data right after it returns to the foreground.

    Will create an issue and fix it in a separate PR.
    
    **UPDATE**: Will take care of it in this issue: https://github.com/wordpress-mobile/WordPress-Android/issues/11377.

![LR_image](https://user-images.githubusercontent.com/1405144/75240389-46340700-57ea-11ea-8a71-f0c074df2bd7.png)

PR submission checklist:

- [X] I have considered adding unit tests where possible.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

malinajirka commented on ImageEditor/src/main/kotlin/org/wordpress/android/imageeditor/preview/PreviewImageFragment.kt:
```diff
@@ -98,4 +105,15 @@ class PreviewImageFragment : Fragment() {
+            UCropUtil.getUCropOptionsBundle(
+                File(inputFilePath),
+                // TODO: output file location
+                File(requireContext().cacheDir, UCropUtil.IMAGE_EDITOR_OUTPUT_IMAGE_FILE_NAME)
```
I think we might want to consider moving this little piece of logic into the VM. We could pass the cacheDir into `vm.onCreateView` and pass both `inputFilePath` and `outputFilePath` into the LiveData. Wdyt?

>UCropUtil.IMAGE_EDITOR_OUTPUT_IMAGE_FILE_NAME

How would you feel about moving this constant into the VM? I feel like it's too specific to live in a utility class, wdyt?

ashiagr replied:
Refactored here 7552d63 and 87fb32a

malinajirka commented on ImageEditor/src/main/kotlin/org/wordpress/android/imageeditor/preview/PreviewImageViewModel.kt:
```diff
@@ -41,7 +44,10 @@ class PreviewImageViewModel : ViewModel() {
-        if (newState == ImageInHighResLoadSuccessUiState && currentState != ImageInHighResLoadSuccessUiState) {
+        if (newState == ImageInHighResLoadSuccessUiState &&
+                currentState != ImageInHighResLoadSuccessUiState &&
+                loadIntoFile.value !is ImageLoadToFileSuccessState
```
I feel like this condition is starting to be difficult to reason about. How would you feel about something like this?

```
val highResImageJustLoadedIntoView = newState != currentState && newState == ImageInHighResLoadSuccessUiState
val imageNotLoadedIntoFile  = loadIntoFile.value !is ImageLoadToFileSuccessState
if (highResImageJustLoadedIntoView && imageNotLoadedIntoFile) {
...
```

ashiagr replied:
Agree. Changed here dd96921

malinajirka commented on ImageEditor/src/main/kotlin/org/wordpress/android/imageeditor/preview/PreviewImageFragment.kt:
```diff
@@ -98,4 +105,15 @@ class PreviewImageFragment : Fragment() {
+            R.id.action_previewFragment_to_ucropFragment,
+            UCropUtil.getUCropOptionsBundle(
+                File(inputFilePath),
+                // TODO: output file location
```
I think the cacheDir is an ok location or do you have any concerns?

ashiagr replied:
No concerns. Just wanted to confirm if it is ok.

ashiagr commented on ImageEditor/src/main/kotlin/org/wordpress/android/imageeditor/preview/PreviewImageFragment.kt:
```diff
@@ -64,6 +67,10 @@ class PreviewImageFragment : Fragment() {
             }
         })
+
+        viewModel.startUCrop.observe(this, Observer { filePath ->
```
Good suggestion! Removed "u" from "uCrop" to generalise it 😀. Let me know if you can think of a better name.

ashiagr commented:
Great review @malinajirka ! I've addressed most of your comments. Answered some here:

> We have a FrameLayout there which draws the toolbar on top (z-axis) of the fragment instead of drawing it above (y-axis). 

Actually that was intentional. I initially planned to not include free style crop rectangle and just provide aspect ratios similar to iOS implementation. As we're not displaying any title on the app bar, so thought of utilising the space behind it by extending the `ImageView` behind it. It could have been benefited us in the landscape mode where we have really less space.

![device-2020-02-26-095712](https://user-images.githubusercontent.com/1405144/75312774-f8b6a900-5880-11ea-85a1-5500466f9c42.png)

Then thought of setting app bar background to `@null `and allow touch events to reach the `ImageView` behind it and have a free style crop rectangle. I can discuss the design with @mbshakti and update it accordingly. 
 
> Is the FrameLayout even necessary? 

Removed, it was an oversight 🤦‍♀️.

> I'd suggest leaving this issue to the end of the project after everything else is done. Wdyt?

Agree

> Imo this workaround is ok, especially since we'll want to have a different behavior when we introduce support for editing multiple images. Wdyt?

Agree

>  I noticed there isn't any "Done" button in the uCrop fragment. Is that something we plan to add?

Yes, I planned to include a check mark on the app bar on the right corner as part of https://github.com/wordpress-mobile/WordPress-Android/issues/11146. That's why I mentioned to investigate more about `popToInclusive` in that task.
I plan to use the `UCropResult` received here: 
https://github.com/wordpress-mobile/WordPress-Android/blob/fa4471a3e8aa70da6ffc91a8a7d9a0052588afdf/ImageEditor/src/main/kotlin/org/wordpress/android/imageeditor/EditImageActivity.kt#L61 I will create a separate task for it.
